netket/variational/classical.py

```diff
@@ -0,0 +1,627 @@
+
+        n_samples_per_node = chain_length * self.sampler.n_chains
+        n_samples = n_samples_per_node * utils.n_nodes
+
+        self._n_samples = n_samples
+        self._chain_length = chain_length
+        self._n_samples_per_node = n_samples_per_node
+        self.reset()
+
+    @property
+    def chain_length(self) -> int:
+        """
+        Length of the markov chain used for sampling configurations.
+
+        If running under MPI, the total samples will be n_nodes * chain_length * n_batches.
+        """
+        return self._chain_length
+
+    @chain_length.setter
+    def chain_length(self, length: int):
+        self.n_samples = length * self.sampler.n_chains * utils.n_nodes
+        self.reset()
+
+    @property
+    def n_discard(self) -> int:
+        """
+        Number of discarded samples at the beginning of the markov chain.
+        """
+        return self._n_discard
+
+    @n_discard.setter
+    def n_discard(self, n_discard: Optional[int]):
+        if n_discard is not None and n_discard < 0:
+            raise ValueError(
+                "Invalid number of discarded samples: n_discard={}".format(n_discard)
+            )
+
+        # don't discard if ExactSampler
+        if isinstance(self.sampler, ExactSampler):
+            if n_discard is not None and n_discard > 0:
+                warnings.warn(
+                    "Exact Sampler does not need to discard samples. Setting n_discard to 0."
+                )
+            n_discard = 0
+
+        self._n_discard = (
+            int(n_discard) if n_discard != None else self.chain_length // 10
+        )
+
+    def reset(self):
+        """
+        Resets the sampled states. This method is called automatically every time
+        that the parameters/state is updated.
+        """
+        self._samples = None
+
+    def sample(
+        self,
+        *,
+        chain_length: Optional[int] = None,
+        n_samples: Optional[int] = None,
+        n_discard: Optional[int] = None,
+    ) -> jnp.ndarray:
+        """
+        Sample a certain number of configurations.
+
+        If one among chain_leength or n_samples is defined, that number of samples
+        are gen erated. Otherwise the value set internally is used.
+
+        Args:
+            chain_length: The length of the markov chains.
+            n_samples: The total number of samples across all MPI ranks.
+            n_discard: Number of discarded samples at the beginning of the markov chain.
+        """
+        if n_samples is None and chain_length is None:
+            chain_length = self.chain_length
+        elif chain_length is None:
+            chain_length = compute_chain_length(self.sampler.n_chains, n_samples)
+
+        if n_discard is None:
+            n_discard = self.n_discard
+
+        self.sampler_state = self.sampler.reset(
+            self._apply_fun, self.variables, self.sampler_state
+        )
```

Comment:
Could we make resetting the chain to random states everytime the mcstate is sampled optional (and call it once in the constructor/sampler setter, since sampler.init_state just returns zeros)
or move this to reset() ?

Reply:
do we reset every time we ask for samples ?

Reply:
do we reset every time we ask for samples ?

Reply:
@gcarleo See  [`mcstate.samples`](https://netket.github.io/netket/nk3/docs/_generated/variational/netket.variational.MCState.html#netket.variational.MCState.samples) vs [`mcstate.sample`](https://netket.github.io/netket/nk3/docs/_generated/variational/netket.variational.MCState.html#netket.variational.MCState.sample). 

If you ask for samples through `.samples` they are cached, and the cache is deleted only if you change parameters/weights or iff you ask explicitly.

The latter is the sampling function, which reset every time you call it.

Reset, does other stuff too. For example it's where the pdf is computed for ExactSampler. Maybe we need a better name.

@inailuig This can be a flag in `MetropolisSampler`.

Reply:
But what I mean is: what do you do after you update the parameters in the neural network during the vmc loop? Do you start the chain from a random state again or do you pick up from the previous value?

Reply:
Ah!
There is now [a flag for that](https://netket.github.io/netket/nk3/docs/_generated/samplers/netket.sampler.MetropolisSampler.html#netket.sampler.MetropolisSampler.reset_chain). Defaults to true (reset the chain)

Reply:
Yes I would default to false in the variational, it's very inefficient to reinitialize to a random value after each step of optimization

Reply:
I can easily set a default for the Metropolis samplers in general.
Doing that for a sampler only when used with a certain optimiser is more ugly.
Should we just turn it on in general, for all applications?